windows/table.cpp

```diff
@@ -0,0 +1,275 @@
+		WC_LISTVIEW,
+		L"",
+		winStyle,
+		hInstance,
+		NULL,
+		TRUE);
+	model->tables.push_back(t);
+	uiWindowsRegisterWM_NOTIFYHandler(t->hwnd, onWM_NOTIFY, uiControl(t));
+	ListView_SetExtendedListViewStyle(t->hwnd, LVS_EX_FULLROWSELECT | LVS_EX_LABELTIP);
+	// TODO: try LVS_EX_AUTOSIZECOLUMNS
+	int n = (*(model->mh->NumRows))(model->mh, model);
+	ListView_SetItemCountEx(t->hwnd, n, 0);
+	return t;
+}
+
+uiTableColumn *uiTableAppendColumn(uiTable *t, const char *name)
+{
+	uiTableColumn *c = uiprivNew(uiTableColumn);
+	c->name = toUTF16(name);
+	c->t = t;
+	c->modelColumn = -1;	// -1 = unassigned
+	// we defer the actual ListView_InsertColumn call until a part is added...
+	t->columns.push_back(c);
+	return c;
+}
+
+void uiTableSetRowBackgroundColorModelColumn(uiTable *t, int modelColumn)
+{
+	// not implemented
+}
+
+static void uiTableDestroy(uiControl *c)
+{
+	uiTable *t = uiTable(c);
+	uiTableModel *model = t->model;
+	std::vector<uiTable*>::iterator it;
+
+	uiWindowsUnregisterWM_NOTIFYHandler(t->hwnd);
+	uiWindowsEnsureDestroyWindow(t->hwnd);
+	// detach table from model
+	for (it = model->tables.begin(); it != model->tables.end(); ++it) {
+		if (*it == t) {
+			model->tables.erase(it);
+			break;
+		}
+	}
+	// free the columns
+	for (auto col: t->columns) {
+		uiprivFree(col->name);
+		uiprivFree(col);
+	}
+	t->columns.~vector<uiTableColumn*>();	// (created with placement new, so just call dtor directly)
+	uiFreeControl(uiControl(t));
+}
+
+static void uiTableMinimumSize(uiWindowsControl *c, int *width, int *height)
+{
+	uiTable *t = uiTable(c);
+	uiWindowsSizing sizing;
+	int x, y;
+
+	// suggested listview sizing from http://msdn.microsoft.com/en-us/library/windows/desktop/dn742486.aspx#sizingandspacing:
+	// "columns widths that avoid truncated data x an integral number of items"
+	// Don't think that'll cut it when some cells have overlong data (eg
+	// stupidly long URLs). So for now, just hardcode a minimum:
+
+	x = 107;	// in line with other controls
+	y = 14*3;	// header + 2 lines (roughly)
+	uiWindowsGetSizing(t->hwnd, &sizing);
+	uiWindowsSizingDlgUnitsToPixels(&sizing, &x, &y);
+	*width = x;
+	*height = y;
+}
+
+static BOOL onWM_NOTIFY(uiControl *c, HWND hwnd, NMHDR *nm, LRESULT *lResult)
+{
+	uiTable *t = uiTable(c);
+	uiTableModelHandler *mh = t->model->mh;
+	BOOL ret = FALSE;
+
+	switch (nm->code) {
+	case LVN_GETDISPINFO:
+	{
+		NMLVDISPINFO* di = (NMLVDISPINFO*)nm;	
+		LVITEM* item = &di->item;
+		if (!(item->mask & LVIF_TEXT)) {
+			break;
+		}
+		int row = item->iItem;
+		int col = item->iSubItem;
+		if (col<0 || col>=(int)t->columns.size()) {
+			break;
+		}
+
+		uiTableColumn *tc = (uiTableColumn*)t->columns[col];
+
+		int mcol = tc->modelColumn;
+		uiTableModelColumnType typ = (*mh->ColumnType)(mh,t->model,mcol);
+		if (typ == uiTableModelColumnString) {
+			void* data = (*(mh->CellValue))(mh, t->model, row, mcol);
+			int n = MultiByteToWideChar(CP_UTF8, 0, (const char*)data, -1, item->pszText, item->cchTextMax);

```

Comment:
This might be a problem. It should be using `toUTF16()`, but I forget what the behavior is if you overwrite the `pszText` pointer with your own... or when it's safe to free. Just mark this as a TODO; I'll probably resolve it a different way.